File: src/DesignPatterns/Behavioural/DIY/Strategy/Strategy.java
package DesignPatterns.Behavioural.DIY.Strategy;

interface EncryptionTool{
    String encrypt(String data);
    String decrypt(String encryptedData);
}

class AESEncryption implements EncryptionTool{

    @Override
    public String encrypt(String data) {
        System.out.println("Encrypting data with AES...");
        return data + " (encrypted with AES) ";
    }

    @Override
    public String decrypt(String encryptedData) {
        System.out.println("Decrypting file using AES...");
        return encryptedData.replace(" (encrypted with AES) ", "");
    }

}

class RSAEncryption implements EncryptionTool{

    @Override
    public String encrypt(String data) {
        System.out.println("Encrypting data with RSA...");
        return data + " (encrypted with RSA) ";
    }

    @Override
    public String decrypt(String encryptedData) {
        System.out.println("Decrypting file using RSA...");
        return encryptedData.replace(" (encrypted with RSA)", "");
    }


}

class UnencryptedData{

    private EncryptionTool tool;
    
    public UnencryptedData(EncryptionTool tool) {
        this.tool = tool;
    }

    public void setencrypt(EncryptionTool tool){
        this.tool = tool;
    }
    public String encrypt(String data){
        return tool.encrypt(data);
    }
    
    public String decrypt(String data){
        return tool.decrypt(data);
    }
}

public class Strategy {
    public static void main(String[] args) {
        
        String unencryptedString = "This string is unencrypted.";
        System.out.println("Unencrypted: "+ unencryptedString);
        System.out.println();

        // encrypted with AES
        UnencryptedData data = new UnencryptedData(new AESEncryption());
        String encryptedWithAES = data.encrypt(unencryptedString);
        System.out.println("Encrypted with AES: " + encryptedWithAES);
        System.out.println();

        // decrypted using AES
        String decryptedUsingAES = data.decrypt(encryptedWithAES);
        System.out.println("Decrypted with AES: " + decryptedUsingAES);
        System.out.println();

        System.out.println("----------------------------------------------------");

        // encrypted with RSA
        data = new UnencryptedData(new RSAEncryption());
        String encryptedWithRSA = data.encrypt(unencryptedString);
        System.out.println("Encrypted with RSA: " + encryptedWithRSA);
        System.out.println();

        // decrypted using RSA
        String decryptedUsingRSA = data.decrypt(encryptedWithRSA);
        System.out.println("Decrypted with RSA: " + decryptedUsingRSA);
        System.out.println();
        
    }

}

File: src/DIY/Strategy/Strategy.ts
// Strategy.ts

interface EncryptionTool {
    encrypt(data: string): string;
    decrypt(encryptedData: string): string;
  }
  
  class AESEncryption implements EncryptionTool {
    encrypt(data: string): string {
      console.log("Encrypting data with AES...");
      return data + " (encrypted with AES) ";
    }
  
    decrypt(encryptedData: string): string {
      console.log("Decrypting file using AES...");
      return encryptedData.replace(" (encrypted with AES) ", "");
    }
  }
  
  class RSAEncryption implements EncryptionTool {
    encrypt(data: string): string {
      console.log("Encrypting data with RSA...");
      return data + " (encrypted with RSA) ";
    }
  
    decrypt(encryptedData: string): string {
      console.log("Decrypting file using RSA...");
      return encryptedData.replace(" (encrypted with RSA) ", "");
    }
  }
  
  class UnencryptedData {
    private tool: EncryptionTool;
  
    constructor(tool: EncryptionTool) {
      this.tool = tool;
    }
  
    public setEncrypt(tool: EncryptionTool): void {
      this.tool = tool;
    }
  
    public encrypt(data: string): string {
      return this.tool.encrypt(data);
    }
  
    public decrypt(data: string): string {
      return this.tool.decrypt(data);
    }
  }
  
  // main
  (function main() {
    const unencryptedString = "This string is unencrypted.";
    console.log("Unencrypted: " + unencryptedString);
    console.log();
  
    // Encrypted with AES
    let data = new UnencryptedData(new AESEncryption());
    const encryptedWithAES = data.encrypt(unencryptedString);
    console.log("Encrypted with AES: " + encryptedWithAES);
    console.log();
  
    // Decrypted using AES
    const decryptedUsingAES = data.decrypt(encryptedWithAES);
    console.log("Decrypted with AES: " + decryptedUsingAES);
    console.log();
  
    console.log("----------------------------------------------------");
  
    // Encrypted with RSA
    data = new UnencryptedData(new RSAEncryption());
    const encryptedWithRSA = data.encrypt(unencryptedString);
    console.log("Encrypted with RSA: " + encryptedWithRSA);
    console.log();
  
    // Decrypted using RSA
    const decryptedUsingRSA = data.decrypt(encryptedWithRSA);
    console.log("Decrypted with RSA: " + decryptedUsingRSA);
    console.log();
  })();
  